### ISSUE TITLE
Memory leak on large files with streaming middleware union 

### ISSUE THREAD

**Issue by zckevin** [indexzero/http-server#138]
I'm trying to serve large files(1 GiB) using http-server, the memory usage is abnormally high, so I got the heap profile and saw lots of 65535 KiB Buffer without being GCed.

![smalloc](https://cloud.githubusercontent.com/assets/666724/6679968/6c4a7888-cc8b-11e4-838b-350551b7649b.png)
![smalloc2](https://cloud.githubusercontent.com/assets/666724/6679971/7dc0dfd0-cc8b-11e4-86c2-06c85cc29d25.png)

But when I switch to connect instead of union, everything is all right then, so I'm pretty sure that something wrong is happening in union.

http-server v0.7.5 
node v0.10.32

**Comment by stianeikeland** [comment]
I'm seeing the same thing here. Requesting/streaming a large file (ex: a movie) will make it consume Gbs of memory. If I request the same file again the memory usage will double. Doesn't seem to be released, in the end the node process is killed by the OS when it runs out of available memory.

**Comment by roccomuso** [comment]
+1 I'm having the same issue when trying to transfer big files from devices with low-memory (like the Raspberry Pi).

Couldn't be related to this the issue? http://stackoverflow.com/questions/15001822/sending-large-image-data-over-http-in-node-js

**Comment by zckevin** [comment]
Followed up: 
Push request https://github.com/flatiron/union/pull/55, but seems this project is not maintained any more, a new dependency should be chosen.

**Comment by indexzero** [comment]
Yes. `union` is not maintained anymore on purpose since it predates `streams2` in Node core. A PR that removes it would be welcome.

**Comment by derhuerst** [comment]
@indexzero

> `union` is not maintained anymore […]. A PR that removes it would be welcome.

I'd be happy to do this. One solution might be to just use Express for the middleware-like behavior. A more lightweight solution might be to used [middl](https://github.com/joakimbeng/middl). What do you think?

**Comment by hdf** [comment]
Problem persists on http-server 0.9.0 and node 6.7.0. Which is too bad, as otherwise this is a great little tool. Have to use HFS instead (http://www.rejetto.com/hfs/?f=dl). :(

**Comment by jcrugzz** [comment]
This should be fixed with in `union@0.4.6`

**Comment by thornjad** [comment]
Is this confirmed fixed? Can this issue be closed?

**Comment by tremby** [comment]
I just tried transferring a huge file after running http-server 0.11.1 with no special command-line arguments and the memory usage stayed reasonable throughout. I don't know whether this answers the question, since there is no example code in this ticket and I don't know if `union` is being used by default.

**Comment by thornjad** [comment]
`union` is still being used (though maybe it shouldn't be), but this issue looks to have been fixed with flatiron/union#54

**Comment by thornjad** [comment]
We will eventually be removing `union` (see #483), so hopefully this issue will go away.
